src/dune_engine/build_system.ml

```diff
@@ -967,49 +971,59 @@ end = struct
       | Install _ -> Memo.Build.return None
       | Context _ -> Source_tree.find_dir sub_dir
     in
-    (* Compute the set of targets and the set of source files that must not be
-       copied *)
-    let source_files_to_ignore, source_dirs_to_ignore =
-      List.fold_left rules ~init:(Path.Build.Set.empty, Path.Build.Set.empty)
-        ~f:(fun (acc_files, acc_dirs) { Rule.targets; mode; loc; _ } ->
-          (* CR-someday amokhov: Remove this limitation. *)
-          let directory_targets_not_supported ~targets =
-            if not (Path.Build.Set.is_empty targets.Targets.Validated.dirs) then
-              User_error.raise ~loc
-                [ Pp.text "Directory targets are not supported for this mode" ]
+    (* Compute the set of sources and targets promoted to the source tree that
+       must not be copied to the build directory. *)
+    let source_files_to_ignore, source_dirnames_to_ignore =
+      List.fold_left rules ~init:(Path.Build.Set.empty, String.Set.empty)
+        ~f:(fun (acc_files, acc_dirnames) { Rule.targets; mode; loc; _ } ->
+          let target_dirnames =
+            Path.Build.Set.to_list_map ~f:Path.Build.basename targets.dirs
+            |> String.Set.of_list
           in
+          (* Check if this rule defines any directory targets that conflict with
+             internal Dune directories listed in [extra_subdirs_to_keep]. *)
```

Comment:
This check could be useful for file targets as well.

Reply:
Good point! I just pushed a test to demonstrate a problem (bad error message) and will fix it in the next commit.

Reply:
Fixed!